File: templates/fsharp/app-mvvm/ViewLocator.fs
namespace AvaloniaAppTemplate

open System
open Avalonia.Controls
open Avalonia.Controls.Templates
open AvaloniaAppTemplate.ViewModels

type ViewLocator() =
    interface IDataTemplate with
        
        member this.Build(data) =
            if isNull data then
                null
            else    
                let name = data.GetType().FullName.Replace("ViewModel", "View", StringComparison.Ordinal)
                let typ = Type.GetType(name)
                if isNull typ then
                    upcast TextBlock(Text = sprintf "Not Found: %s" name)
                else
                    let view = Activator.CreateInstance(typ) :?> Control
                    view.DataContext <- data
                    view
                
        member this.Match(data) = data :? ViewModelBase

File: templates/csharp/app-mvvm/ViewLocator.cs
using System;
#if (!ReactiveUIToolkitChosen)
using System.ComponentModel;
#endif
using Avalonia.Controls;
using Avalonia.Controls.Templates;
#if (ReactiveUIToolkitChosen)
using AvaloniaAppTemplate.ViewModels;
#endif

namespace AvaloniaAppTemplate
{
    public class ViewLocator : IDataTemplate
    {
        public IControl Build(object data)
        {
            var name = data.GetType().FullName!.Replace("ViewModel", "View");
            var type = Type.GetType(name);

            if (type != null)
            {
                return (Control)Activator.CreateInstance(type)!;
            }
            
            return new TextBlock { Text = "Not Found: " + name };
        }

        public bool Match(object data)
        {

#if (ReactiveUIToolkitChosen)
            return data is ViewModelBase;
#else
            return data is INotifyPropertyChanged;
#endif
        }
    }
}